src/System.Buffers/src/System/Buffers/ArrayPoolBucket.cs

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+
+namespace System.Buffers
+{
+    /// <summary>
+    /// Provides a thread-safe bucket containing buffers that can be Rented and Returned as part 
+    /// of a buffer pool; it should not be used independent of the pool.
+    /// </summary>
+    internal sealed class ArrayPoolBucket<T>
+    {
+        private volatile int _index;
+        private T[][] _data;
+        private SpinLock _lock;
+
+        /// <summary>
+        /// Creates the pool with numberOfBuffers arrays where each buffer is of bufferLength length.
+        /// </summary>
+        internal ArrayPoolBucket(int bufferLength, int numberOfBuffers)
+        {
+            _index = 0;
+            _lock = new SpinLock();
+            _data = new T[numberOfBuffers][];
+            for (int i = 0; i < numberOfBuffers; i++)
+            {
+                _data[i] = new T[bufferLength];
+            }
```

Comment:
I appreciate your reasoning.  However...

> The configuration options will allow them to fine tune how many buffers they need

That's assuming they're creating their own pool. Which also assumes they're not using any other components (e.g. corefx itself) that's using the shared pool.

>  all calls to Rent (except when the bucket is empty) are constant time

This again seems to assume not using the shared buffer.  If a developer is just calling ```ArrayPool<byte>.Shared.Rent(value)``` everything is happening as part of the call to Rent, and it's not visible whether the cost is the allocation of the pool or the allocation of the bucket or the allocation of the array.  Further, if that value is coming from some non-hardcoded value, the developer can't have expectations that the bucket has been initialized.

Imagine we use this within System.IO.BufferedStream to get the buffer used to back the stream.  In an ASP.NET app, there's a good chance we'll quickly hit the max limit from the bucket, at which point it'll have the characteristic you want.  But imagine now we're in a WinRT app.  It's much less likely there a developer is going to be doing a lot of parallelism, yet they may still use one BufferedStream at a time, with a default buffer size of 80K, which for the default pool would mean they've now increased their working set instantly by ~6MB even though they really only used 80K at a time.

I think it's important that we not forcefully incur more memory overheads than necessary.  At some point we probably even want to explore throwing away buffers even if we're under the limit.